### PR TITLE
TOFTOF GUI added testing automation

### LINKED ISSUE #22791: Automated testing for TOFTOF GUI
For the moment, TOFTOF GUI is not covered by the automated tests at all. A set of unit tests should be created.

We will appreciate any advice or recommendations.

Milistone: 3.14

@JoachimCoenen will work on it.

### PR DESCRIPTION
**Pull request by JoachimCoenen**
**Description of work.**
added `TOFTOFGUITest.py` and `TOFTOFScriptElementTest.py` under `scripts/test/TOFTOF/`

**To test:**

`TOFTOFGUITest.py` and `TOFTOFScriptElementTest.py` tests get executed and pass without failure.

Fixes #22791. 

Does this update require release notes?
- [ ] Yes
- [ ] No

Milestone: 3.14

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by mganeva** [comment]
Label: In Progress. I need some time to test it.

File: scripts/test/TOFTOF/TOFTOFGUITest.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+class TOFTOFScriptElementTest(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls):
+        cls.app = QtGui.QApplication([])
+
+    def setUp(self):
+        self.settingsMock = mock.Mock()
+        facility_name = "name of facility"
+        self.settingsMock.facility_name = facility_name
```
Comment (soininen) [inline scripts/test/TOFTOF/TOFTOFGUITest.py]: Why not `self.settingsMock.facility_name = "name of facility"`? Ditto for instrument name below.

File: scripts/test/TOFTOF/TOFTOFGUITest.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+        instrument_name = "name of instrument"
+        self.settingsMock.instrument_name = instrument_name
+        self.setupWidget = TOFTOFSetupWidget(self.settingsMock)
+
+    def tearDown(self):
+        self.settingsMock = None
+        self.setupWidget = None
+        #pass
+        #self.scriptElement = None
+        #done
```
Comment (soininen) [inline scripts/test/TOFTOF/TOFTOFGUITest.py]: Commented-out code?

File: scripts/test/TOFTOF/TOFTOFGUITest.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+    #def test_that_init_works_correctly(self):
```
Comment (soininen) [inline scripts/test/TOFTOF/TOFTOFGUITest.py]: Commented-out code?

File: scripts/test/TOFTOF/TOFTOFGUITest.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
+from __future__ import (absolute_import, division, print_function)
+
+from reduction_gui.reduction.toftof.toftof_reduction import TOFTOFScriptElement, OptionalFloat
+from reduction_gui.widgets.toftof.toftof_setup import TOFTOFSetupWidget
+import mantid
```
Comment (soininen) [inline scripts/test/TOFTOF/TOFTOFGUITest.py]: Please, run `flake8` on the test files: it says, for example, that `mantid` is an unused import.

File: scripts/test/TOFTOF/TOFTOFGUITest.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+        # empty can runs, comment, and factor
+        scriptElement.ecRuns   = 'TOFTOFTestdata.nxs'
+        scriptElement.ecTemp   = OptionalFloat(21.0)
+        scriptElement.ecFactor = 0.9
+
+        # data runs: [(runs,comment, temperature), ...]
+        scriptElement.dataRuns = [
+            [u'TOFTOFTestdata.nxs', u'H2O 21C', OptionalFloat(None)],
+            [u'TOFTOFTestdata.nxs', u'H2O 34C', OptionalFloat(34.0)]
```
Comment (soininen) [inline scripts/test/TOFTOF/TOFTOFGUITest.py]: Just a note: my understanding is that the `u` prefix for strings is a syntax error in early Python 3 versions (3.0, 3.2?). I hope when Mantid finally moves to Python 3 everyone will be running a later version...

File: scripts/test/TOFTOF/TOFTOFScriptElementTest.py [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+        self.scriptElement.keepSteps     = True
+
+        # save data
+        self.scriptElement.saveDir      = ''
+        self.scriptElement.saveSofQW    = False
+        self.scriptElement.saveSofTW    = False
+        self.scriptElement.saveNXSPE    = False
+        self.scriptElement.saveNexus    = False
+        self.scriptElement.saveAscii    = False
+        exec('from mantid.simpleapi import *\n' + self.scriptElement.to_script(), dict(), dict())
```
Comment (soininen) [inline scripts/test/TOFTOF/TOFTOFScriptElementTest.py]: There don't seem to be any asserts in this test. What is it testing then?

File: scripts/test/TOFTOF/TOFTOFScriptElementTest.py [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+        ({'vanRuns': '0:2', 'vanCmnt': ''}, True),
+        ({'vanRuns': '0:2', 'vanCmnt': 'Comment for Vanadium'}, False),
+
+        ({'saveNXSPE': True, 'saveSofQW': True,  'saveDir': ''}, True),
+        ({'saveNXSPE': True, 'saveSofQW': False, 'saveDir': ''}, True),
+        ({'saveNXSPE': True, 'saveSofQW': False, 'saveDir': '/some/SaveDir/'}, True),
+        ({'saveNXSPE': True, 'saveSofQW': True,  'saveDir': '/some/SaveDir/'}, False),
+        ]
+
+        def executeSubTest(inputs, shoudThrow):
```
Comment (soininen) [inline scripts/test/TOFTOF/TOFTOFScriptElementTest.py]: `shoudThrow` -> `shouldThrow`?

**Review by soininen — CHANGES_REQUESTED** [review]
Good to have some unit tests for the TOFTOF interface. The tests pass and the changes look generally good. Below are a few minor comments.

**Comment by soininen** [comment]
We have moved to `clang-format 5.0`, please merge master to this branch. And a small note for the future: please explain the changes in the commit messages, 'fixing requested changes' does not really tell what the commit is about.

**Comment by JoachimCoenen** [comment]
ok.

**Review by soininen — CHANGES_REQUESTED** [review]

File: scripts/test/TOFTOF/TOFTOFScriptElementTest.py [hunk 8]
```diff
@@ -181,8 +167,12 @@ def test_that_script_is_executable_in_mantid(self):
...
         self.scriptElement.keepSteps     = True
 
-        with self.assertRaisesNothing("Generated reduction script cannot be executed."):
-            exec('from mantid.simpleapi import *\n' + self.scriptElement.to_script(), dict(), dict())
+        def execScript(script):
+            exec(script, dict(), dict())
+
+        testhelpers.assertRaisesNothing(self, execScript, 'from mantid.simpleapi import *\n' + self.scriptElement.to_script())
+        #with self.assertRaisesNothing("Generated reduction script cannot be executed."):
+        #    exec('from mantid.simpleapi import *\n' + self.scriptElement.to_script(), dict(), dict())
```
Comment (soininen) [inline scripts/test/TOFTOF/TOFTOFScriptElementTest.py]: Commented-out code?

**Review by soininen — APPROVED** [review]

**Pull request merged**